### ISSUE TITLE
Non-deterministic segfault in prterun from base/rmaps_base_binding.c

### ISSUE THREAD

**Issue by lrbison** [openpmix/prrte#1573]
## Background information

I am testing openmpi-v5.0.x branch (2b638cfa6989b8712fb4138eec251b8c174fbcb6), but am observing non-deterministic failures on process launch.  Core dumps occur in PRRTE, but I'm unsure if this is a prrte problem, a PMIx problem, or an OpenMPI problem.

### What version of the PMIx Reference RTE (PRRTE) are you using? (e.g., v2.0, v3.0, git master @ hash, etc.)
Version ab03675e5a9014418418555ceb188d2573713870 (just after v3.0.0rc3 tag)
##### What version of PMIx are you using? (e.g., v4.2.0, git branch name and hash, etc.)
Version b22eca2c462b61533572634a0abbf212f283578d (just after v4.2.2rc2 tag)
### Please describe the system on which you are running
* Operating system/version: AL2
* Computer hardware: P3dn
* Network type: EFA & TCP

-----------------------------

## Details of the problem

Please describe, in detail, the problem that you are having, including the behavior you expect to see, the actual behavior that you are seeing, steps to reproduce the problem, etc.  It is most helpful if you can attach a small program that a developer can use to reproduce your problem.

I'm running a simple bash script which just echoes the hostname.  I'm not sure the executable matters as I think the crash happens before the processes are launched, but it doesn't happen all the time, I observe the crash only after about 15-20 runs:
```shell
$ mpirun -n 2 --hostfile hostfile hostname
```

I can capture the core dump and this is the stack I see:
```
#0  hwloc_bitmap_and (res=0x8399f0, set1=0x0, set2=0x861aa0) at bitmap.c:1064
#1  0x00007fe10389b53c in bind_generic (jdata=0x84ebc0, proc=0x850e60, node=0x847fc0, obj=0x0, options=0x7ffee2fb4560) at base/rmaps_base_binding.c:96
#2  0x00007fe10389bea6 in prte_rmaps_base_bind_proc (jdata=0x84ebc0, proc=0x850e60, node=0x847fc0, obj=0x0, options=0x7ffee2fb4560)
    at base/rmaps_base_binding.c:371
#3  0x00007fe10389781e in prte_rmaps_base_setup_proc (jdata=0x84ebc0, idx=0, node=0x847fc0, obj=0x0, options=0x7ffee2fb4560)
    at base/rmaps_base_support_fns.c:615
#4  0x00007fe1038aa6cf in ppr_mapper (jdata=0x84ebc0, options=0x7ffee2fb4560) at rmaps_ppr.c:275
#5  0x00007fe103891897 in prte_rmaps_base_map_job (fd=-1, args=4, cbdata=0x8419c0) at base/rmaps_base_map_job.c:655
#6  0x00007fe1026da3ad in event_process_active_single_queue (activeq=0x764d60, base=0x764870) at event.c:1350
#7  event_process_active (base=<optimized out>) at event.c:1420
#8  event_base_loop (base=0x764870, flags=1) at event.c:1621
#9  0x00000000004080fd in main (argc=16, argv=0x7ffee2fb5458) at prte.c:1077
```

Via `ldd` I can see that I'm using my system's hwloc (/lib64/hwloc.so.5).
As you can tell from the stack hwloc_bitmap_and is being called with set1 NULL which causes the segfault.  It looks like the rest of the node structure in [rmaps_base_bind.c#L96](https://github.com/openpmix/prrte/blob/ab03675e5a9014418418555ceb188d2573713870/src/mca/rmaps/base/rmaps_base_binding.c#L96) is sensible as far as I can tell.

prrte was configured as follows:
```shell
./configure --disable-option-checking --prefix=/fsx/libraries/openmpi/v5.0.x/install --enable-prte-ft --with-proxy-version-string=5.0.0rc9 --with-proxy-package-name=Open MPI --with-proxy-bugreport=https://www.open-mpi.org/community/help/ --enable-prte-prefix-by-default --disable-pmix-lib-checks --with-pmix-extra-libs=/fsx/libraries/openmpi/v5.0.x/source/ompi/3rd-party/openpmix/src/libpmix.la --enable-debug --without-verbs --with-libfabric=/opt/amazon/efa --disable-man-pages --enable-cuda --with-cuda=/usr/local/cuda --with-cuda-libdir=/usr/local/cuda/lib64/stubs --disable-builtin-atomics --enable-debug CPPFLAGS= -I/fsx/libraries/openmpi/v5.0.x/source/ompi/3rd-party/openpmix/include -I/fsx/libraries/openmpi/v5.0.x/source/ompi/3rd-party/openpmix/include -I/fsx/libraries/openmpi/v5.0.x/source/ompi/3rd-party/openpmix/ -I/fsx/libraries/openmpi/v5.0.x/source/ompi/3rd-party/openpmix/ --cache-file=/dev/null --srcdir=.
```

At this point I'm out of my depth so any suggestions appreciated.

**Comment by rhc54** [comment]
Just to be clear - there is no DVM involved here, correct? This is just repetitive independent executions of `mpirun`?

**Comment by lrbison** [comment]
Yes, you are correct, no DVM involved, independent `mpirun` executions will trigger the problem.

**Comment by rhc54** [comment]
Okay - is this is a multi-node operation? Looking at the code, it appears that the `prte_node_t` object we use to track the node did not get properly initialized.

**Comment by lrbison** [comment]
The hostfile has two nodes.  By including a `--map-by node` it will run one rank on each, but if I omit the mapping, it'll choose to run both ranks on the first node.  Both of these mapping configurations can result in a segfault in the same location.

**Comment by rhc54** [comment]
I'm trying to understand something about what you are doing - are you building OMPI using the internal PRRTE and PMIx code? I'm trying to figure out why you are providing a PRRTE configure line here - are you implying you used an external version of PRRTE, but pointed at the PMIx internal to OMPI?

**Comment by lrbison** [comment]
Yes, I'm trying to use OMPI defaults from the unreleased v5.0.x branch.  I included the PRRTE configuration from the open-mpi "3rd party" directory in case it helped, but PRRTE was configured by the openmpi configure and openmpi chose the commit id for me.

I have been playing with using both internal and external hwloc from the openmpi configure options with no change.

**Comment by rhc54** [comment]
I just wanted to ensure you weren't attempting to mix internal and external versions of these packages as that won't work (which is why the info about external hwloc and prrte config lines was so confusing). This isn't a configuration issue (assuming you do build from all internal code drops). I'm just trying to figure out what is different as we test the "run lots of mpiruns" constantly, without problems.

**Comment by rhc54** [comment]
FWIW: I have no idea what an "AL2" or "P3dn" are - but from the "EFA" reference, I gather you are running on some AWS instance?

**Comment by rhc54** [comment]
@wckzhang Can someone over there take a look at this? We cycle hundreds of `mpirun` executions in multiple environments without issue, so this must be one of those things that gets tickled by the AWS-environment again.

**Comment by rhc54** [comment]
I'm not sure if this is the issue or not, but you might give https://github.com/openpmix/prrte/pull/1574 a try and see if that fixes it.

**Comment by rhc54** [comment]
@lrbison FWIW: our testing system doesn't have a head-node with a different topology - our test systems are all homogeneous. You are technically operating in a heterogeneous environment, hence the difference.

**Comment by lrbison** [comment]
Thank you!  This does seem to fix the segfault.  While I was checking, the OpenMPI v5 branch was updated so I have now checked the fix on both OpenMPI's 2b638cfa6989b8712fb4138eec251b8c174fbcb6 as well as the newer ecdb3ba798b6098a6b6388d1faaf1087dab43f1f.  

Furthermore, after getting into a working condition I manually disabled the change you added and immediately the issue re-appeared.

Lastly without your fix I launched the run from one of the compute nodes so it would be a homogeneous cluster, and as expected no error occurs, confirming the problem only occurs in a heterogeneous cluster.

Thank you for the help.  OK to close.
